### PR TITLE
feat(backingimage): backing image clone and encryption support

### PR DESCRIPTION
**Pull request by ChanYiLin**
ref: https://github.com/longhorn/longhorn/issues/7051

Check the LEP here: https://github.com/longhorn/longhorn/pull/8740
This PR is depends on the
- backing-image-manager PR: https://github.com/longhorn/backing-image-manager/pull/251

### CODE REVIEW

File: controller/backing_image_controller.go [hunk 1]
```diff
@@ -401,6 +401,20 @@ func (bic *BackingImageController) handleBackingImageDataSource(bi *longhorn.Bac
 			if err != nil {
 				return err
 			}
+
+			// For cloning, we choose the same node and disk as the source backing image
```
Comment (derekbit) [inline controller/backing_image_controller.go:405]: nit: `handleBackingImageDataSource` is too longhorn to read easily. Suggest refactoring the codes in the future.

**Comment by ChanYiLin** [comment]
Feedback from UI, thanks to @a110605 
- webhook should check if the secret exists or not
- create backing image / restore backing image should allow user to specify the secret if the backing image is already encrypted
- clone an encrypted backing image should inherit the secret

File: webhook/resources/backingimage/validator.go [hunk 2]
```diff
@@ -107,3 +119,27 @@ func (b *backingImageValidator) Delete(request *admission.Request, oldObj runtim
...
+
+	if bimtypes.EncryptionType(targetBackingImage.Spec.SourceParameters[longhorn.DataSourceTypeCloneParameterEncryption]) == bimtypes.EncryptionTypeEncrypt ||
+		bimtypes.EncryptionType(targetBackingImage.Spec.SourceParameters[longhorn.DataSourceTypeCloneParameterEncryption]) == bimtypes.EncryptionTypeDecrypt {
+		if targetBackingImage.Spec.SourceParameters[longhorn.DataSourceTypeCloneParameterSecret] == "" ||
+			targetBackingImage.Spec.SourceParameters[longhorn.DataSourceTypeCloneParameterSecretNamespace] == "" {
+			return werror.NewInvalidError("secret is not provided for encryption or decryption during cloneing", "")
+		}
+	}
+
+	return nil
```
Comment (derekbit) [inline webhook/resources/backingimage/validator.go:144]: nit:
```
encryptionType := bimtypes.EncryptionType(targetBackingImage.Spec.SourceParameters[longhorn.DataSourceTypeCloneParameterEncryption])

if encryptionType == bimtypes.EncryptionTypeEncrypt {
    ....
}

...
```

File: controller/backing_image_controller.go [hunk 3]
```diff
@@ -401,6 +401,20 @@ func (bic *BackingImageController) handleBackingImageDataSource(bi *longhorn.Bac
...
+			if bi.Spec.SourceType == longhorn.BackingImageDataSourceTypeClone {
+				sourceBackingImageName := bi.Spec.SourceParameters[longhorn.DataSourceTypeCloneParameterBackingImage]
+				sourceBackingImage, err := bic.ds.GetBackingImageRO(sourceBackingImageName)
+				if err != nil {
+					return fmt.Errorf("failed to get source backing image %v during cloning", sourceBackingImageName)
+				}
+				readyNode, readyDiskName, err = bic.ds.GetOneBackingImageReadyNodeDisk(sourceBackingImage)
+				if err != nil {
+					return fmt.Errorf("failed to find one ready source backing image %v during cloning", sourceBackingImageName)
+				}
```
Comment (derekbit) [inline controller/backing_image_controller.go:415]: Move the codes to a helper function, e.g.

```
func findReadyNodeAndDisk(...) (nodeName, diskName string, err error) {
    ...
}
```

File: controller/backing_image_data_source_controller.go [hunk 4]
```diff
@@ -669,7 +682,11 @@ func (c *BackingImageDataSourceController) generateBackingImageDataSourcePodMani
 		"--source-type", string(bids.Spec.SourceType),
 	}
 
-	if err := c.prepareRunningParameters(bids); err != nil {
+	bids.Status.RunningParameters = bids.Spec.Parameters
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply controller/backing_image_data_source_controller.go:685]: I think it is because besides the parameters provided by the users 
we have some parameters derived after
this `RunningParameters` is originally for `export-from-volume`
the parameter users provided for `export-from-volume` is `Volume Name`
But we derived others parameters for the execution
```
	DataSourceTypeExportFromVolumeParameterVolumeName                = "volume-name"    // from users directly

// derived after
	DataSourceTypeExportFromVolumeParameterVolumeSize                = "volume-size"
	DataSourceTypeExportFromVolumeParameterSnapshotName              = "snapshot-name"
	DataSourceTypeExportFromVolumeParameterSenderAddress             = "sender-address"
	DataSourceTypeExportFromVolumeParameterFileSyncHTTPClientTimeout = "file-sync-http-client-timeout"
```
You can see the function `prepareRunningParametersForExport()`


Same reason, we get the backing image UUID for running parameters inside the function
because it is inconvenient for users to fill in the UUID when cloning

**Comment by ChanYiLin** [comment]
> Feedback from UI, thanks to @a110605
> 
> * webhook should check if the secret exists or not
> * create backing image / restore backing image should allow user to specify the secret if the backing image is already encrypted
> * clone an encrypted backing image should inherit the secret

Have addressed the changes

File: controller/backing_image_data_source_controller.go [hunk 5]
```diff
@@ -405,6 +405,19 @@ func (c *BackingImageDataSourceController) syncBackingImage(bids *longhorn.Backi
 		}
 	}
 
+	// Only copy the secret to spec if it is to encrypt other backing image
+	// because we use spec secret to check if it is encrypted.
+	if bi.Spec.SourceType == longhorn.BackingImageDataSourceTypeClone &&
+		bimtypes.EncryptionType(bi.Spec.SourceParameters[longhorn.DataSourceTypeCloneParameterEncryption]) == bimtypes.EncryptionTypeEncrypt {
+
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply controller/backing_image_data_source_controller.go:412]: fixed.

File: controller/backing_image_data_source_controller.go [hunk 6]
```diff
@@ -679,6 +696,23 @@ func (c *BackingImageDataSourceController) generateBackingImageDataSourcePodMani
 		cmd = append(cmd, "--checksum", bids.Spec.Checksum)
 	}
 
+	if bids.Spec.SourceType == longhorn.BackingImageDataSourceTypeClone &&
+		bids.Spec.Parameters[longhorn.DataSourceTypeCloneParameterSecretNamespace] != "" &&
+		bids.Spec.Parameters[longhorn.DataSourceTypeCloneParameterSecret] != "" {
+
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply controller/backing_image_data_source_controller.go:702]: fixed.

File: webhook/resources/backingimage/mutator.go [hunk 7]
```diff
@@ -114,11 +114,30 @@ func (b *backingImageMutator) Create(request *admission.Request, newObj runtime.
...
+			return nil, werror.NewInvalidError(err.Error(), "")
+		}
+	}
+
+	if oldBackingImage.Spec.SecretNamespace != "" {
+		if oldBackingImage.Spec.SecretNamespace != backingImage.Spec.SecretNamespace {
+			err := fmt.Errorf("changing secret namespace for BackingImage %v is not supported", oldBackingImage.Name)
+			return nil, werror.NewInvalidError(err.Error(), "")
+		}
+	}
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply webhook/resources/backingimage/mutator.go:154]: fixed, thanks

File: webhook/resources/backingimage/validator.go (hunk 2 above)
Comment (ChanYiLin) [reply webhook/resources/backingimage/validator.go]: fixed.

File: controller/backing_image_controller.go (hunk 3 above)
Comment (ChanYiLin) [reply controller/backing_image_controller.go]: fixed.

**Comment by ChanYiLin** [comment]
> This pull request is now in conflict. Could you fix it @ChanYiLin? 🙏

fixed.

File: k8s/generate_code.sh [hunk 8]
```diff
@@ -37,14 +37,14 @@ if [[ ! -d "${GOPATH}/src/k8s.io/code-generator" ]]; then
 fi
 
 # https://github.com/kubernetes-sigs/controller-tools/tree/v0.7.0/cmd/controller-gen
-if ! command -v ${GOPATH}/bin/controller-gen > /dev/null; then
```
Comment (ChanYiLin) [inline k8s/generate_code.sh:40]: this is to revert the change from my previous PR
this change is only for my own env

File: webhook/resources/backingimage/mutator.go [hunk 9]
```diff
@@ -91,6 +93,19 @@ func (b *backingImageMutator) Create(request *admission.Request, newObj runtime.
...
+			sourceBackingImageName := parameters[longhorn.DataSourceTypeCloneParameterBackingImage]
+			sourceBackingImage, err := b.ds.GetBackingImageRO(sourceBackingImageName)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to get source backing image %v", sourceBackingImageName)
+			}
+			patchOps = append(patchOps, fmt.Sprintf(`{"op": "replace", "path": "/spec/secret", "value": "%s"}`, sourceBackingImage.Spec.Secret))
+			patchOps = append(patchOps, fmt.Sprintf(`{"op": "replace", "path": "/spec/secretNamespace", "value": "%s"}`, sourceBackingImage.Spec.SecretNamespace))
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply webhook/resources/backingimage/mutator.go]: fixed.

File: webhook/resources/backingimage/mutator.go [hunk 10]
```diff
@@ -91,6 +93,27 @@ func (b *backingImageMutator) Create(request *admission.Request, newObj runtime.
...
+			sourceBackingImageName := parameters[longhorn.DataSourceTypeCloneParameterBackingImage]
+			sourceBackingImage, err := b.ds.GetBackingImageRO(sourceBackingImageName)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to get source backing image %v", sourceBackingImageName)
+			}
+			patchOps = append(patchOps, fmt.Sprintf(`{"op": "replace", "path": "/spec/secret", "value": "%s"}`, sourceBackingImage.Spec.Secret))
+			patchOps = append(patchOps, fmt.Sprintf(`{"op": "replace", "path": "/spec/secretNamespace", "value": "%s"}`, sourceBackingImage.Spec.SecretNamespace))
+			if sourceBackingImage.Status.Checksum == "" {
+				return nil, errors.Wrapf(err, "failed to get checksum of source backing image %v", sourceBackingImageName)
+			}
```
Comment (derekbit) [inline webhook/resources/backingimage/mutator.go:108]: Do we need the check if checksum is empty?

Comment (ChanYiLin) [reply webhook/resources/backingimage/mutator.go:108]: I think we can always override the checksum no matter if it is empty

**Comment by ChanYiLin** [comment]
> This pull request is now in conflict. Could you fix it @ChanYiLin? 🙏

fixed.

**Review by derekbit — APPROVED** [review]
LGTM

**Review by derekbit — APPROVED** [review]
